crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -258,58 +258,59 @@ impl CumulocityConverter {
-                payload: mqtt_payload.chars().take(50).collect(),
-            }
-        })?;
-        let child_id = tedge_event.source.clone();
-        let need_registration = if let Some(child_id) = child_id.clone() {
-            add_external_device_registration_message(child_id, &mut self.children, &mut messages)
-        } else {
-            false
+        let event_type = match event_type.is_empty() {
+            true => DEFAULT_EVENT_TYPE,
+            false => event_type,
         };
 
-        let c8y_event = C8yCreateEvent::try_from(tedge_event)?;
+        if let Some(entity) = self.entity_store.get(source) {
+            let mqtt_topic = input.topic.name.clone();
+            let mqtt_payload = input.payload_str().map_err(|e| {
+                ThinEdgeJsonDeserializerError::FailedToParsePayloadToString {
+                    topic: mqtt_topic.clone(),
+                    error: e.to_string(),
+                }
+            })?;
+
+            let tedge_event =
+                ThinEdgeEvent::try_from(event_type, entity, mqtt_payload).map_err(|e| {
+                    ThinEdgeJsonDeserializerError::FailedToParseJsonPayload {
+                        topic: mqtt_topic.clone(),
+                        error: e.to_string(),
+                        payload: mqtt_payload.chars().take(50).collect(),
+                    }
+                })?;
 
-        // If the message doesn't contain any fields other than `text` and `time`, convert to SmartREST
-        let message = if c8y_event.extras.is_empty() {
-            let smartrest_event = Self::serialize_to_smartrest(&c8y_event)?;
-            let smartrest_topic = Topic::new_unchecked(SMARTREST_PUBLISH_TOPIC);
-            Message::new(&smartrest_topic, smartrest_event)
-        } else {
-            // If the message contains extra fields other than `text` and `time`, convert to Cumulocity JSON
-            let cumulocity_event_json = serde_json::to_string(&c8y_event)?;
-            let json_mqtt_topic = Topic::new_unchecked(C8Y_JSON_MQTT_EVENTS_TOPIC);
-            Message::new(&json_mqtt_topic, cumulocity_event_json)
+            let c8y_event = C8yCreateEvent::try_from(tedge_event)?;
+
+            // If the message doesn't contain any fields other than `text` and `time`, convert to SmartREST
+            let message = if c8y_event.extras.is_empty() {
+                let smartrest_event = Self::serialize_to_smartrest(&c8y_event)?;
+                let smartrest_topic = Topic::new_unchecked(SMARTREST_PUBLISH_TOPIC);
+                Message::new(&smartrest_topic, smartrest_event)
+            } else {
+                // If the message contains extra fields other than `text` and `time`, convert to Cumulocity JSON
+                let cumulocity_event_json = serde_json::to_string(&c8y_event)?;
+                let json_mqtt_topic = Topic::new_unchecked(C8Y_JSON_MQTT_EVENTS_TOPIC);
+                Message::new(&json_mqtt_topic, cumulocity_event_json)
+            };
```

Comment:
We agreed offline to simplify this by systematically publishing the events over the c8y json topic. This will be done in a follow-up PR, though.